### PR TITLE
Add VLM-aware SpecPrefill to batched media routes

### LINKED ISSUE #506: Design SpecPrefill support for media-bearing MLLMScheduler routes
## Problem\n\nSpecPrefill works locally on Qwen 27B/35B TextModel-direct routes, but media-bearing VLM requests route through MLLMScheduler / MLLMBatchGenerator and do not engage SpecPrefill. The current MLLM path has no draft-model/specprefill wiring.\n\nBlindly sparse-skipping text tokens inside a media prefill path is unsafe because Qwen VLM requests depend on image/video placeholders, media embeddings, and media-aware MRoPE position deltas.\n\n## Local evidence\n\nRuntime route-parity artifact:\n\n- /opt/ai-runtime/run/qwen-feature-parity/20260505T-route-recovery/route-recovery-summary.md\n\nObserved locally on Qwen 27B/35B:\n\n- TextModel-direct SpecPrefill: engaged and produced valid output.\n- Media VLM under SpecPrefill-enabled profile: request passed through mllm_batch, output was valid, but specprefill_engaged=false.\n\n## Design requirement\n\nA VLM-aware SpecPrefill design should compose with the existing media prefill path instead of treating media prompts as plain text. It should account for:\n\n- get_input_embeddings / vision encoding\n- image/video placeholder tokens\n- media-aware MRoPE position IDs and deltas\n- prefix-cache composition\n- request-local logits processors\n- EOS and stop-token behavior\n- fallback when sparse prefill is unsafe\n- per-request enable/disable behavior\n\n## Acceptance criteria\n\n- TextModel-direct SpecPrefill remains valid.\n- Media baseline remains valid with SpecPrefill off.\n- Media SpecPrefill either engages safely with nonempty valid output or explicitly reports why it is unavailable for the route/model.\n- Prefix cache and SpecPrefill composition are tested on media-bearing requests.\n- The design avoids local-only route drift between TextModel-direct and MLLMScheduler behavior.

### LINKED ISSUE #511: Design VLM-aware SpecPrefill for media MLLMScheduler routes
# Design VLM-aware SpecPrefill for media MLLMScheduler routes

## Summary

Text-only SpecPrefill can run through TextModel-direct paths, but media-bearing
VLM requests route through `MLLMScheduler` and `MLLMBatchGenerator`. That route
needs VLM-aware SpecPrefill wiring before media prompts can use SpecPrefill
without dropping media-derived inputs or position/mask state.

## Local repro

Local Runtime evidence:

- `/opt/ai-runtime/run/live-feature-parity/20260506T-media-specprefill-safetyfloor-27b/summary.json`
- `/opt/ai-runtime/run/live-feature-parity/20260506T-media-specprefill-safetyfloor-35b/summary.json`
- `/opt/ai-runtime/run/live-feature-parity/20260506T-122b-closure-board-specprefill/run/qwen3.5-122b-a10b/09-media-specprefill/route-media-vlm.json`

Observed before local media-route wiring:

- the baseline media request completed
- `specprefill_engaged=false`
- text-only paths could engage SpecPrefill
- the gap was route parity for media-bearing requests, not global SpecPrefill
  absence

## Exact upstream code path compared

Repository: `waybarrios/vllm-mlx`

Commit compared: `1e24e156f52c26bf96d55d605f266d9debcfa801`

Paths/functions:

- `vllm_mlx/mllm_batch_generator.py::MLLMBatchGenerator`
- `vllm_mlx/mllm_scheduler.py::MLLMScheduler`
- `vllm_mlx/engine/batched.py::BatchedEngine`
- `vllm_mlx/specprefill.py`

The upstream route has `MLLMScheduler`, `MLLMBatchGenerator`, and
`specprefill.py`, but the matching media-aware SpecPrefill surface is not
present: no `_media_specprefill_enabled`, no `_run_media_specprefill`, and no
`sparse_prefill_inputs_embeds` path that preserves media-derived embeddings and
visual position masks through selection.

## Expected behavior

When a media-bearing VLM request is eligible and SpecPrefill is enabled, the
MLLM route should run a media-aware sparse prefill path and return generation
metadata with `specprefill_engaged=true`, while preserving nonempty final
content, position handling, visual masks, logits processors, EOS behavior, and
fallback semantics.

## Minimal patch/design ask

- Add an explicit media SpecPrefill eligibility point in `MLLMBatchGenerator`.
- Add a VLM-aware sparse prefill helper that accepts media-derived input
  embeddings and visual position masks.
- Preserve fallback to the baseline media path when eligibility or prefill
  execution is not safe.
- Add focused tests for media inputs, metadata, EOS behavior, and fallback.

## What this does not claim

- This does not claim SpecPrefill is safe for every VLM architecture.
- This does not claim media SpecPrefill should be enabled by default.
- This does not claim text-only SpecPrefill behavior is broken upstream.

### PR DESCRIPTION
**Pull request by waybarrios**
## Summary

Closes #506.
Closes #511.

Media-bearing requests use `MLLMScheduler` and `MLLMBatchGenerator`, while the existing SpecPrefill implementation only covered TextModel-direct routes. This PR adds an opt-in VLM-aware SpecPrefill path for batched Qwen media requests.

Supported runtime identities:

- Qwen3-VL dense
- Qwen3-VL MoE
- Qwen3.5 dense
- Qwen3.5 MoE

Unknown or mismatched implementations fail closed and continue through the existing dense media prefill path.

### Media-aware sparse prefill

The new `vllm_mlx/mllm_specprefill.py` adapter obtains the model’s fused media state through `get_input_embeddings` before selecting prompt chunks:

```python
features = model.get_input_embeddings(
    input_ids=input_ids,
    pixel_values=request.pixel_values,
    **kwargs,
)
```

The adapter selects the same positions across all request-local media state:

```python
selected_tokens = features.input_ids[:, index_array]
selected_embeds = features.inputs_embeds[:, index_array, :]
selected_positions = features.position_ids[..., index_array]
selected_visual_mask = features.visual_pos_masks[..., index_array]
```

Selection always retains:

- image and video placeholder chunks
- visual start and end boundary chunks
- the final prompt token
- the first uncached chunk when composing a cached text prefix

Deep-stack visual embeddings are gathered using visual-token ordinals, so selected visual state stays aligned with its placeholders.

### Request-local MRoPE handling

The implementation does not install or mutate shared RoPE wrappers.

Instead, each request carries its own adjusted decode delta:

```python
compression_adjustment = original_count - selected_count
decode_rope_delta = features.rope_deltas + compression_adjustment
```

`MLLMBatch` keeps these deltas aligned when rows are created, filtered, or extended. Decode calls pass the current batch values explicitly:

```python
model_kwargs["rope_deltas"] = self.active_batch.decode_rope_deltas
output = self.language_model(input_tokens, cache=cache, **model_kwargs)
```

This prevents one media request from leaking mutable MRoPE state into another row during continuous batching.

### Safe prefix-cache composition

A media request may reuse a text prefix only when the cached portion ends before the visual start marker:

```python
if remaining_ids and can_compose_text_prefix_cache(
    request.input_ids,
    model.config,
    cached_count,
):
    media_prefix_for_specprefill = True
```

The uncached suffix still contains the full visual span and is processed through the media-aware adapter.

Qwen3.5 hybrid `ArraysCache` containers are cloned before reuse, including their mutable state list and batching metadata:

```python
new_cache = ArraysCache(len(cache.cache))
new_cache.cache = list(cache.cache)
new_cache.left_padding = cache.left_padding
new_cache.lengths = cache.lengths
```

This prevents a media suffix from modifying the recurrent state stored under an earlier text-prefix key.

Sparse media caches are not stored as reusable token-only prefix entries.

### Draft-model compatibility

Media placeholder and visual-boundary IDs may be outside the draft model’s vocabulary. Only those known media IDs are replaced for importance scoring. Any unrelated out-of-vocabulary token rejects sparse scoring and uses the dense fallback.

Qwen3.5 drafts receive the gated-query extractor explicitly:

```python
if draft_model_type in {"qwen3_5", "qwen3_5_moe"}:
    score_kwargs["query_extractor"] = _qwen35_extract_queries
```

Qwen3-VL drafts retain the standard Qwen attention extractor selected by `score_tokens`.

### Eligibility and fallbacks

Server defaults and request-level overrides are represented by:

```python
SpecPrefillRequestConfig(
    enabled=...,
    keep_pct=...,
    backbone_pct=...,
)
```

The sparse route fails closed for:

- unsupported or mismatched model identities
- missing or incompatible draft models
- audio inputs
- prompts below the configured threshold
- prompts above the scoring safety cap
- invalid keep or backbone percentages
- MTP execution
- interleaved chunked prefill
- rotating caches
- non-sparse selections
- runtime scoring or prefill failures

All these cases use a fresh dense media cache so a partial sparse attempt cannot contaminate the fallback.

### Diagnostics

Each request receives a `SpecPrefillOutcome` containing:

- whether SpecPrefill was requested
- whether it engaged
- the decision or fallback reason
- target model identity
- original, selected, and cached token counts

The final OpenAI response exposes this information through `generation_metadata`:

```json
{
  "specprefill_requested": true,
  "specprefill_engaged": true,
  "specprefill_reason": "engaged",
  "specprefill_route": "mllm_media",
  "specprefill_original_tokens": 9000,
  "specprefill_selected_tokens": 3100,
  "specprefill_cached_tokens": 256
}
```

Streaming responses preserve the metadata even when a reasoning or tool parser suppresses the final text delta. When necessary, a terminal metadata-only chunk is emitted before `[DONE]`.

Scheduler statistics also report aggregate decisions, fallback counts, selected tokens, and selection rate.

SpecPrefill remains disabled by default. Existing TextModel-direct behavior is unchanged.

## Type of change

- New feature
- Performance
- Documentation

## Surface touched

- CLI / serve command
- OpenAI API endpoints
- Scheduler / batching
- KV cache / prefix cache
- Multimodal pipeline

## Test plan

Linux-compatible tests use synthetic Qwen runtime identities and NumPy-backed media state. Apple Silicon CI runs the MLX-dependent server, cache, scheduler, and continuous-batching coverage.

```bash
pytest -q \
  tests/test_mllm_media_specprefill.py \
  tests/test_batched_engine_mllm_config.py \
  tests/test_request.py \
  tests/test_api_models.py \
  tests/test_batched_engine.py \
  tests/test_server.py

ruff check vllm_mlx/ tests/ \
  --select E,F,W \
  --ignore E402,E501,E731,F811,F841

black --check vllm_mlx/ tests/

python -m py_compile \
  vllm_mlx/mllm_specprefill.py \
  vllm_mlx/mllm_batch_generator.py \
  vllm_mlx/mllm_scheduler.py \
  vllm_mlx/engine/batched.py \
  vllm_mlx/server.py

git diff --check
```

CI now includes:

- `tests/test_mllm_media_specprefill.py`
- `tests/test_batched_engine_mllm_config.py`
- Apple Silicon prefix-cache coverage for hybrid `ArraysCache` isolation

Regression coverage includes:

- exact Qwen dense and MoE capability identities
- unsupported and mismatched model fallback
- image, video, boundary, and final-token retention
- fused embedding and MRoPE position gathering
- deep-stack visual embedding alignment
- prefix-cache composition boundaries
- hybrid recurrent-cache isolation
- request-level controls and threshold behavior
- audio and runtime incompatibility fallbacks
- draft vocabulary safety
- Qwen3.5 and Qwen3-VL draft extractor selection
- cancellation
- engine and scheduler configuration forwarding
- API diagnostics and suppressed-final-delta streaming metadata

## Test results

- Focused suite: `171 passed, 123 skipped, 3 deselected`
- Media SpecPrefill and engine wiring: `21 passed`
- Ruff: pass
- Black: pass
- Python compilation: pass
- `git diff --check`: pass
- Independent frozen-diff review: clean

## Performance impact

Eligible long media prompts prefill only selected target-model chunks while retaining the complete required visual state. The draft model still scores the full prompt.

No Apple Silicon benchmark was run locally, so this PR does not claim a measured TTFT or throughput improvement. The feature remains opt-in and uses dense prefill whenever sparse execution is unsafe or unavailable.

## Output parity

With SpecPrefill disabled, below threshold, or unavailable, requests continue through the existing dense media path.

When SpecPrefill engages, byte-for-byte parity is not expected because sparse prefill intentionally removes low-importance text chunks. The implementation preserves media embeddings, visual masks, deep-stack state, request-local MRoPE positions, logits processors, sampling parameters, EOS handling, stop tokens, and the final prompt token.

Apple Silicon output validation is delegated to CI because MLX cannot run in the local Linux environment.

## Risk notes

- Support uses an exact runtime allowlist and fails closed for unknown implementations.
- Audio, MTP, interleaved chunked prefill, and rotating caches use explicit dense fallbacks.
- Shared model RoPE modules are never patched.
- Decode deltas remain request-local throughout continuous batching.
- Media prefix-cache composition is restricted to prefixes ending before the visual span.
- Hybrid cache containers and mutable state lists are cloned before reuse.
- Dense fallback starts with a fresh cache after any sparse attempt.
- Sparse media caches are not stored as token-only reusable prefixes.
- API additions are optional diagnostics and do not remove or rename existing response fields.

### CODE REVIEW

**Comment by waybarrios** [comment]
Updated this PR onto the current main and resolved the overlapping generator, scheduler, engine, server, MTP/MRoPE, and prefix-cache changes. Added regression coverage for media SpecPrefill, cache isolation, audio routing, configuration, and streaming metadata. Local Ruff, Black, py_compile, and the available unit tests pass; Apple/MLX inference remains for CI validation.
